make server_urls dynamically reloadable

kananindzya opened the pull request:
<!--

Replace this comment with a description of what's being changed by this PR.

If this PR should close an issue, please add one of the magic keywords
(e.g. Fixes) followed by the issue number. For more info see:
https://help.github.com/articles/closing-issues-using-keywords/

If this pull request is work in progress, create a draft PR instead of prefixing the title with WIP.

-->
closes elastic/apm-agent-java#723 
### Checklist

<!-- Potential tasks related to a new PR. Remove tasks that are not relevant -->

- [x] Implement code
- [x] Add tests
- [ ] Update documentation
- [x] Update [CHANGELOG.md](CHANGELOG.md)
- [x] Update [supported-technologies.asciidoc](docs/supported-technologies.asciidoc)

kananindzya commented:
> Unfortunately, it's not as easy as that. The `ApmServerClient#serverUrls` variable has to be updated as well, in a thread-safe manner. Meaning that the variable should be declared volatile.
> 
> See also [#725 (comment)](https://github.com/elastic/apm-agent-java/pull/725#issuecomment-511819759) for more details on how to use change listeners on how to register a config change callback. You can do that in the constructor of `ApmServerClient`. Also, reset the error count when the config changes.



> Also, reset the err

I added listener, but not in ApmServerClient constructor. And added
```
    public static synchronized void overrideApmServerUrls(List<URL> serverUrls) {
        logger.debug("server_urls override with value = ({}).", serverUrls);
        ApmServerClient.serverUrls = serverUrls;
    }
```
but I don't like this decision.

kananindzya commented:
```
2019-07-19 16:34:43.076 [apm-reporter] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Receiving TRANSACTION event (sequence 1)
2019-07-19 16:34:43.077 [apm-reporter] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Starting new request to http://localhost:8201/intake/v2/events
2019-07-19 16:34:43.077 [apm-reporter] ERROR co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Failed to handle event of type TRANSACTION with this error: Connection refused (Connection refused)
2019-07-19 16:34:43.077 [apm-reporter] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Event handling failure
java.net.ConnectException: Connection refused (Connection refused)
        at java.net.PlainSocketImpl.socketConnect(Native Method)
        at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
        at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
        at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
        at java.net.Socket.connect(Socket.java:589)
        at sun.net.NetworkClient.doConnect(NetworkClient.java:175)
        at sun.net.www.http.HttpClient.openServer(HttpClient.java:463)
        at sun.net.www.http.HttpClient.openServer(HttpClient.java:558)
        at sun.net.www.http.HttpClient.<init>(HttpClient.java:242)
        at sun.net.www.http.HttpClient.New(HttpClient.java:339)
        at sun.net.www.http.HttpClient.New(HttpClient.java:357)
        at sun.net.www.protocol.http.HttpURLConnection.getNewHttpClient(HttpURLConnection.java:1220)
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1156)
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1050)
        at sun.net.www.protocol.http.HttpURLConnection.connect(HttpURLConnection.java:984)
        at co.elastic.apm.agent.report.IntakeV2ReportingEventHandler.startRequest(IntakeV2ReportingEventHandler.java:200)
        at co.elastic.apm.agent.report.IntakeV2ReportingEventHandler.handleEvent(IntakeV2ReportingEventHandler.java:138)
        at co.elastic.apm.agent.report.IntakeV2ReportingEventHandler.onEvent(IntakeV2ReportingEventHandler.java:116)
        at co.elastic.apm.agent.report.IntakeV2ReportingEventHandler.onEvent(IntakeV2ReportingEventHandler.java:50)
        at co.elastic.apm.agent.shaded.lmax.disruptor.BatchEventProcessor.processEvents(BatchEventProcessor.java:168)
        at co.elastic.apm.agent.shaded.lmax.disruptor.BatchEventProcessor.run(BatchEventProcessor.java:125)
        at java.lang.Thread.run(Thread.java:748)
2019-07-19 16:34:43.078 [apm-reporter] INFO co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Backing off for 1 seconds (+/-10%)
2019-07-19 16:34:50.673 [configuration-reloader] DEBUG co.elastic.apm.agent.report.ApmServerClient - server_urls override with value = [[http://localhost:8200]].
2019-07-19 16:34:50.747 [apm-reporter] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Receiving METRICS event (sequence 2)
2019-07-19 16:34:50.748 [apm-reporter] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Starting new request to http://localhost:8200/intake/v2/events
2019-07-19 16:34:50.758 [apm-reporter] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Scheduling request timeout in 10s
2019-07-19 16:35:00.758 [apm-request-timeout-timer] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Request flush because the request timeout occurred
2019-07-19 16:35:00.759 [apm-reporter] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Receiving FLUSH event (sequence 3)
2019-07-19 16:35:00.761 [apm-reporter] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Flushing 1381 uncompressed 596 compressed bytes
^C2019-07-19 16:35:17.759 [apm-reporter] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Receiving SHUTDOWN event (sequence 4)
```

felixbarny commented on apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java:
```diff
@@ -73,6 +74,16 @@ public ApmServerClient(ReporterConfiguration reporterConfiguration) {
 
     public ApmServerClient(ReporterConfiguration reporterConfiguration, List<URL> serverUrls) {
         this.reporterConfiguration = reporterConfiguration;
+        for (ConfigurationOption configurationOption : reporterConfiguration.getConfigurationOptions()) {
```
Better create a method `public ConfigurationOption<List<URL>> getServerUrlsOption()` in `ReporterConfiguration` and then just call `reporterConfiguration.getServerUrlsOption().addChangeListener(...)` here

kananindzya replied:
fixed

felixbarny commented on apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java:
```diff
@@ -245,4 +256,10 @@ int getErrorCount() {
         T withConnection(HttpURLConnection connection) throws IOException;
     }
 
+    public synchronized void overrideApmServerUrls(List<URL> serverUrls) {
```
Just inline this to the `onChange` method. Doesn't have to be synchronized.

kananindzya replied:
fixed

felixbarny commented on apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java:
```diff
@@ -245,4 +256,10 @@ int getErrorCount() {
 
+    public synchronized void overrideApmServerUrls(List<URL> serverUrls) {
+        logger.debug("server_urls override with value = ({}).", serverUrls);
+        this.serverUrls = serverUrls;
```
Shuffle the URLs by reusing the `shuffleUrls` method. To achieve that, change the parameter of `shuffleUrls` from `ReporterConfiguration` to `List<URL>`.

kananindzya replied:
added

kananindzya commented:
@felixbarny  I fixed according to comments

felixbarny commented:
could you add a section to the changelog?

kananindzya commented:
> could you add a section to the changelog?

added

felixbarny commented:
Easier than using an actual properties file for testing is to use an in-memory PropertySource like here:
https://github.com/elastic/apm-agent-java/blob/45857cea69a82ac97d5f1bee266f58fbedddb43e/apm-agent-plugins/apm-slf4j-plugin/src/test/java/co/elastic/apm/agent/slf4j/Slf4JMdcActivationListenerTest.java#L175

eyalkoren commented on apm-agent-core/src/test/java/co/elastic/apm/agent/report/ApmServerClientTest.java:
```diff
@@ -24,18 +24,30 @@
 import java.net.URLConnection;
+import java.util.Arrays;
 import java.util.List;
+import java.util.ServiceLoader;
```
remove unused imports

kananindzya replied:
removed

eyalkoren commented on apm-agent-core/src/test/java/co/elastic/apm/agent/report/ApmServerClientTest.java:
```diff
@@ -52,17 +64,27 @@
     @Rule
     public WireMockRule apmServer2 = new WireMockRule(WireMockConfiguration.wireMockConfig().dynamicPort());
     private ApmServerClient apmServerClient;
+    protected static ConfigurationRegistry config;
```
```suggestion
    private ConfigurationRegistry config;
```

eyalkoren commented on apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java:
```diff
@@ -240,6 +251,10 @@ int getErrorCount() {
         return errorCount.get();
     }
 
+    protected List<URL> getServerUrls()  {
```
```suggestion
   List<URL> getServerUrls()  {
```

kananindzya commented:
> `getServerUrls`

hi @eyalkoren , I added protected ApmServerClient#getServerUrls, and added test with using
```
config.save("server_urls", tempUrl.toString(), SpyConfiguration.CONFIG_SOURCE_NAME);
```

felixbarny commented:
can you merge in the changes from master again? that should fix the test failures

felixbarny commented:
run the tests
